File: agent/src/main/java/pt/um/lei/masb/agent/net/JSONMappedURLs.java
package pt.um.lei.masb.agent.net;

import pt.um.lei.masb.agent.data.DataSource;
import pt.um.lei.masb.agent.data.apis.ApiSource;
import pt.um.lei.masb.blockchain.data.Category;

import java.util.List;
import java.util.Map;

public class JSONMappedURLs {
    private List<ApiSource> apis;
    private Map<Category, DataSource[]> tcp;
    private DataSource jade[];

    JSONMappedURLs() {
    }

    public List<ApiSource> getApis() {
        return apis;
    }

    public Map<Category, DataSource[]> getTcp() {
        return tcp;
    }

    public DataSource[] getTemperatureCategory() {
        return tcp.getOrDefault(Category.TEMPERATURE, new DataSource[0]);
    }

    public DataSource[] getHumidityCategory() {
        return tcp.getOrDefault(Category.HUMIDITY, new DataSource[0]);
    }

    public DataSource[] getLuminosityCategory() {
        return tcp.getOrDefault(Category.LUMINOSITY, new DataSource[0]);
    }

    public DataSource[] getNoiseCategory() {
        return tcp.getOrDefault(Category.NOISE, new DataSource[0]);
    }

    public DataSource[] getOtherCategory() {
        return tcp.getOrDefault(Category.OTHER, new DataSource[0]);
    }

    public DataSource[] getJade() {
        return jade;
    }

}

File: agent/src/main/kotlin/pt/um/lei/masb/agent/net/JSONMappedURLs.kt
package pt.um.lei.masb.agent.net

import kotlinx.serialization.Serializable
import mu.KLogging
import pt.um.lei.masb.agent.Container
import pt.um.lei.masb.agent.data.DataSource
import java.io.File
import java.io.FileReader
import java.io.InputStreamReader
import java.net.URI


@Serializable
data class JSONMappedURLs(
    val apis: List<DataSource> = emptyList(),
    val tcp: List<DataSource> = emptyList(),
    val jade: List<DataSource> = emptyList()
) {
    companion object : KLogging() {
        val apiScheme = Regex.fromLiteral("http(s)?")
        val tcpScheme = Regex.fromLiteral("tcp")
        val jadeScheme = Regex.fromLiteral("jade")

        /**
         * A parser for a file based on a path string from the classpath resources.
         *
         * @param file the path string to a resource in the classpath.
         */
        fun jsonFileUrlParser(file: String): JSONMappedURLs =
            InputStreamReader(
                Container::class.java
                    .getResourceAsStream(file)
            ).use {
                //JSON.parse(this.serializer(), it.readText())
                TODO()
            }

        /**
         * A parser for a file at the specified URI.
         *
         * @param file
         */
        fun jsonFileUrlParser(file: URI): JSONMappedURLs =
            FileReader(File(file)).use {
                //JSON.parse(this.serializer(), it.readText())
                TODO()
            }


    }


}
